hoodie-spark/src/test/java/HoodieJavaApp.java

```diff
@@ -0,0 +1,148 @@
+        HoodieTestDataGenerator dataGen = new HoodieTestDataGenerator();
+
+        /**
+         * Commit with only inserts
+         */
+        // Generate some input..
+        List<String> records1 = DataSourceTestUtils.convertToStringList(dataGen.generateInserts("001"/* ignore */, 100));
+        Dataset<Row> inputDF1 = spark.read().json(jssc.parallelize(records1, 2));
+
+        // Save as hoodie dataset (copy on write)
+        inputDF1.write()
+                .format("com.uber.hoodie") // specify the hoodie source
+                .option("hoodie.insert.shuffle.parallelism", "2") // any hoodie client config can be passed like this
+                .option("hoodie.upsert.shuffle.parallelism", "2") // full list in HoodieWriteConfig & its package
+                .option(DataSourceWriteOptions.OPERATION_OPT_KEY(), DataSourceWriteOptions.INSERT_OPERATION_OPT_VAL()) // insert
+                .option(DataSourceWriteOptions.RECORDKEY_FIELD_OPT_KEY(), "_row_key") // This is the record key
+                .option(DataSourceWriteOptions.PARTITIONPATH_FIELD_OPT_KEY(), "partition") // this is the partition to place it into
+                .option(DataSourceWriteOptions.PRECOMBINE_FIELD_OPT_KEY(), "timestamp") // use to combine duplicate records in input/with disk val
+                .option(HoodieWriteConfig.TABLE_NAME, tableName) // Used by hive sync and queries
+                .mode(SaveMode.Overwrite) // This will remove any existing data at path below, and create a new dataset if needed
+                .save(tablePath); // ultimately where the dataset will be placed
+        String commitInstantTime1 = HoodieDataSourceHelpers.latestCommit(fs, tablePath);
+
+        /**
+         * Commit that updates records
+         */
+        List<String> records2 = DataSourceTestUtils.convertToStringList(dataGen.generateUpdates("002"/* ignore */, 100));
+        Dataset<Row> inputDF2 = spark.read().json(jssc.parallelize(records2, 2));
+        inputDF2.write()
+                .format("com.uber.hoodie")
+                .option("hoodie.insert.shuffle.parallelism", "2")
+                .option("hoodie.upsert.shuffle.parallelism", "2")
+                .option(DataSourceWriteOptions.RECORDKEY_FIELD_OPT_KEY(), "_row_key")
+                .option(DataSourceWriteOptions.PARTITIONPATH_FIELD_OPT_KEY(), "partition")
+                .option(DataSourceWriteOptions.PRECOMBINE_FIELD_OPT_KEY(), "timestamp")
+                .option(HoodieWriteConfig.TABLE_NAME, tableName)
+                .mode(SaveMode.Append)
+                .save(tablePath);
+        String commitInstantTime2 = HoodieDataSourceHelpers.latestCommit(fs, tablePath);
+
+        /**
+         * Read & do some queries
+         */
+        Dataset<Row> hoodieROViewDF = spark.read()
+                .format("com.uber.hoodie")
+                // pass any path glob, can include hoodie & non-hoodie datasets
+                .load(tablePath + "/*/*/*/*");
```

Comment:
Why pass a pattern to glob? Do we not know this from partition metadata?

Reply:
Because `spark.read.parquet` takes a glob of paths. Idea is to support a mix of hoodie/non-hoodie paths.. Underneath, we do look up partition metadata to check if this is a hoodie dataset and so forth.